src/irmin-pack/unix/control_file_intf.ml

```diff
@@ -0,0 +1,164 @@
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+open! Import
+
+module Payload_v3 = struct
+  type from_v1_v2_post_upgrade = { entry_offset_at_upgrade_to_v3 : int63 }
+  [@@deriving irmin]
+  (** [entry_offset_at_upgrade_to_v3] is the offset of the first entry that is
+      known to have been created using [irmin_pack_version = `V2] or more. The
+      entries before that point may be v1 entries. V1 entries need an entry in
+      index because it is the only place their lenght is stored. *)
+
+  type from_v3_gc_allowed = {
+    entry_offset_suffix_start : int63;
+    gc_generation : int;
+  }
+  [@@deriving irmin]
+  (** [entry_offset_suffix_start] is 0 if the suffix file was never garbage
+      collected. Otherwise it is the offset of the very first entry of the
+      suffix file. Note that offsets in the suffix file are virtual. The garbage
+      collections don't reset the offsets.
+
+      [gc_generation] is the number of past GCs. A suffix file and a prefix file
+      containing that integer in their name exist. *)
+
+  (** [From_v1_v2_post_upgrade] corresponds to a pack store that was upgraded to
+      [`V3]. It contains infos related to backward compatibility. GCs are
+      forbidden on it.
+
+      TODO: Change to something better for future upgrades.
+
+      [from_v3_gc_disallowed] corresponds to a pack store that was created using
+      [`V3] code. GCs are forbidden on it because it uses an indexing strategy.
+
+      [from_v3_gc_allowed] corresponds to a pack store that was created using
+      [`V3] code. GCs are possible on it, but the indexing strategy is
+      constrained to minimal. *)
+  type status =
+    | From_v1_v2_post_upgrade of from_v1_v2_post_upgrade
+    | From_v3_gc_disallowed
+    | From_v3_gc_allowed of from_v3_gc_allowed
+  [@@deriving irmin]
+
+  type t = {
+    dict_offset_end : int63;
+    entry_offset_suffix_end : int63;
+    status : status;
+  }
+  [@@deriving irmin]
+  (** Payload of the control file for [irmin_pack_version = `V3].
+
+      [dict_offset_end] is the offset in the dict file just after the last valid
+      dict bytes. The next data to be pushed to the dict will be pushed at this
+      offset.
+
+      [entry_offset_suffix_end] is similar to [dict_offset_end] but for the
+      suffix file. *)
+end
+
+module Latest_payload = Payload_v3
+
+module type S = sig
+  module Io : Io.S
+
+  type t
+
+  val create_rw :
+    path:string ->
+    overwrite:bool ->
+    use_fsync:bool ->
+    Latest_payload.t ->
+    (t, [> Io.create_error | Io.write_error ]) result
+  (** Create a rw instance of [t] by creating a control file. *)
```

Comment:
Here and elsewhere I think we need to document what `use_fsync` is for: presumably for correctness we must use fsync, but for speed (accepting that there may be correctness issues) we can opt to not use fsync?

Reply:
I've moved the fsyncs to a single location in flush. Documentation will be needed there